src/Lexer.js

```diff
@@ -49,11 +49,16 @@ function mangle(text) {
  */
 module.exports = class Lexer {
   constructor(options) {
+    this.setOptions(options);
+  }
+
+  setOptions(options) {
     this.tokens = [];
     this.tokens.links = Object.create(null);
     this.options = options || defaults;
-    this.options.tokenizer = this.options.tokenizer || new Tokenizer();
-    this.tokenizer = this.options.tokenizer;
+    if (!this.tokenizer) {
```

Comment:
If they pass a different tokenizer in the options wouldn't we want to use that one here?

Reply:
That still happens, if my logic is correct.

Instead of this with every loop:
1) Check for `this.options.tokenizer`
2) If it doesn't exist, assign it the default value `new Tokenizer`
3) Assign the result to `this.tokenizer`

Do this:
1) Check for `this.tokenizer`
2) If it doesn't exist, assign it a value, preferably `options.tokenizer`, but if not, default to `new Tokenizer`

The end result is the same: `this.tokenizer == options.tokenizer || new Token`, but we have saved an object write in the second case.

Instead of "saving" the optional Tokenizer to `this.options.tokenizer` to be reapplied every loop, we "save" it where it actually belongs, in `this.tokenizer`, and only have to apply it the first time around.

Reply:
In fact... I wonder if we even need this separate "setOptions" function run every time. Once the tokenizer is defined and its options are set the first time, normal use would generally not be swapping options on the fly, right? Wouldn't 99% of use cases just set options one time and then run `marked(string)` with those same options every time?

The only real reason I see the need to assign options on the fly is for unit testing in development, because it needs to update options as it steps through the spec tests for GFM, Pedantic, etc. in a random order.

Reply:
What happens in this situation?

```js
marked("# test");

marked.use({
  tokenizer: {
    heading: () {
      // no heading
    }
  }
});

marked("# test");
```

with this PR `this.tokenizer` wouldn't be updated and the new heading tokenizer is ignored.